pkg/kubelet/kubelet.go

```diff
@@ -699,11 +699,19 @@ func (kl *Kubelet) initialNodeStatus() (*api.Node, error) {
 		}
 		// TODO(roberthbailey): Can we do this without having credentials to talk
 		// to the cloud provider?
-		instanceID, err := instances.ExternalID(kl.hostname)
+		externalID, err := instances.ExternalID(kl.hostname)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get external ID from cloud provider: %v", err)
+		}
+		node.Spec.ExternalID = externalID
+		// TODO: We can't assume that the node has credentials to talk to the
+		// cloudprovider from arbitrary nodes. At most, we should talk to a
+		// local metadata server here.
+		instanceID, err := instances.InstanceID(node.Name)
 		if err != nil {
 			return nil, fmt.Errorf("failed to get instance ID from cloud provider: %v", err)
 		}
-		node.Spec.ExternalID = instanceID
+		node.Spec.ProviderID = kl.cloud.ProviderName() + "://" + instanceID
 	} else {
 		node.Spec.ExternalID = kl.hostname
```

Comment:
@roberthbailey @bgrant0607 I was trying to be more consistent than `ExternalID`: if there's no provider then there's no `ProviderID`. I was preferring an empty string rather than forging a `"none://" + kl.hostname` uri. In fact `node.Spec.ExternalID = kl.hostname` is also part of the original problem that I am trying to solve with this PR. So unless you really think I should do it I'd skip this suggestion. When we'll have a "baremetal" provider then this will be automatically filled it with the proper name and instance id.